dask/dataframe/io/parquet/arrow.py

```diff
@@ -335,9 +335,13 @@ def read_metadata(
         ignore_metadata_file=False,
         metadata_task_size=0,
         parquet_file_extension=None,
+        storage_options=None,
         **kwargs,
     ):
 
+        # Optimize open_file_func for remote filesystems
+        cls._update_open_file_func(fs, storage_options, kwargs)
```

Comment:
Could we make this something more like:

```python
kwargs = cls._update_open_file_func(fs, storage_options, kwargs)
```

or (psuedocode below):

```python
if "open_file_options" not in kwargs:
    open_file_options = cls._get_open_file_options(fs, storage_options, kwargs)
    if open_file_options:
        kwargs["open_file_options"] = open_file_options
```

Where possible I'd like to avoid methods that mutate things inplace as it can make it more difficult to reason about the logic at play. Being more explicit about where things are set is, I think, a good thing.

Reply:
>Where possible I'd like to avoid methods that mutate things inplace as it can make it more difficult to reason about the logic at play. Being more explicit about where things are set is, I think, a good thing.

I completely agree that code should be functional whenever possible. Sorry, you were so quick to review that I didn't get a chance to clean this up